Consolidate contiguous memory copies

ekuznetsov139 opened the pull request:
I see a lot of framework overhead with batch matmul operations, because the stream executor generates a large number of hipMemcpyD2DAsync calls for each batch matmul:

> 
> .[32m<<hip-api pid:18001 tid:3.34015 hipMemcpyDtoDAsync (0x7fa804587800, 0x7fa803e00000, 4096, stream:0.1) @12376130256341.[0m
> .[32m<<hip-api pid:18001 tid:3.34016 hipMemcpyDtoDAsync (0x7fa804588800, 0x7fa803e01000, 4096, stream:0.1) @12376130278984.[0m
> .[32m<<hip-api pid:18001 tid:3.34017 hipMemcpyDtoDAsync (0x7fa804589800, 0x7fa803e02000, 4096, stream:0.1) @12376130291748.[0m
> .[32m<<hip-api pid:18001 tid:3.34018 hipMemcpyDtoDAsync (0x7fa80458a800, 0x7fa803e03000, 4096, stream:0.1) @12376130303700.[0m
> .[32m<<hip-api pid:18001 tid:3.34019 hipMemcpyDtoDAsync (0x7fa80458b800, 0x7fa803e04000, 4096, stream:0.1) @12376130315483.[0m
> .[32m<<hip-api pid:18001 tid:3.34020 hipMemcpyDtoDAsync (0x7fa80458c800, 0x7fa803e05000, 4096, stream:0.1) @12376130326984.[0m
> .[32m<<hip-api pid:18001 tid:3.34021 hipMemcpyDtoDAsync (0x7fa80458d800, 0x7fa803e06000, 4096, stream:0.1) @12376130338747.[0m
> .[32m<<hip-api pid:18001 tid:3.34022 hipMemcpyDtoDAsync (0x7fa80458e800, 0x7fa803e07000, 4096, stream:0.1) @12376130350308.[0m
> .[32m<<hip-api pid:18001 tid:3.34023 hipMemcpyDtoDAsync (0x7fa80458f800, 0x7fa803e00000, 4096, stream:0.1) @12376130361840.[0m
> .[32m<<hip-api pid:18001 tid:3.34024 hipMemcpyDtoDAsync (0x7fa804590800, 0x7fa803e01000, 4096, stream:0.1) @12376130374163.[0m
> .[32m<<hip-api pid:18001 tid:3.34025 hipMemcpyDtoDAsync (0x7fa804591800, 0x7fa803e02000, 4096, stream:0.1) @12376130385795.[0m
> .[32m<<hip-api pid:18001 tid:3.34026 hipMemcpyDtoDAsync (0x7fa804592800, 0x7fa803e03000, 4096, stream:0.1) @12376130397487.[0m
> .[32m<<hip-api pid:18001 tid:3.34027 hipMemcpyDtoDAsync (0x7fa804593800, 0x7fa803e04000, 4096, stream:0.1) @12376130409119.[0m
> .[32m<<hip-api pid:18001 tid:3.34028 hipMemcpyDtoDAsync (0x7fa804594800, 0x7fa803e05000, 4096, stream:0.1) @12376130420801.[0m
> (snip)
> 
> .[32m<<hip-api pid:18001 tid:3.35016 hipMemcpyDtoDAsync (0x7fa804970800, 0x7fa803e01000, 4096, stream:0.1) @12376144032051.[0m
> .[32m<<hip-api pid:18001 tid:3.35017 hipMemcpyDtoDAsync (0x7fa804971800, 0x7fa803e02000, 4096, stream:0.1) @12376144042461.[0m
> .[32m<<hip-api pid:18001 tid:3.35018 hipMemcpyDtoDAsync (0x7fa804972800, 0x7fa803e03000, 4096, stream:0.1) @12376144052740.[0m
> .[32m<<hip-api pid:18001 tid:3.35019 hipMemcpyDtoDAsync (0x7fa804973800, 0x7fa803e04000, 4096, stream:0.1) @12376144066035.[0m
> .[32m<<hip-api pid:18001 tid:3.35020 hipMemcpyDtoDAsync (0x7fa804974800, 0x7fa803e05000, 4096, stream:0.1) @12376144076194.[0m
> .[32m<<hip-api pid:18001 tid:3.35021 hipMemcpyDtoDAsync (0x7fa804975800, 0x7fa803e06000, 4096, stream:0.1) @12376144087786.[0m
> .[32m<<hip-api pid:18001 tid:3.35022 hipMemcpyDtoDAsync (0x7fa804976800, 0x7fa803e07000, 4096, stream:0.1) @12376144097935.[0m
> .[32m<<hip-api pid:18001 tid:3.35026 18001 3.35026 hipLaunchKernel 'Cijk_Ailk_Bjlk_SB_MT32x32x2_SE_AF0EM1_AMAS2_ASEM1_BL0_ETSP_EPS0_GRCGA1_GRCGB1_GRVW4_GSU1_ISA0_0_0_IU1_K1_KLS_LPB0_MGWVW1_NLCA1_NLCB1_PGR1_PLR1_SGR1_SIA1_SLW1_SU32_SUM0_TT4_4_USFGRO0_VA
> W1_VS1_VW4_WG8_8_1_WGM1' gridDim:{1,1,1008} groupDim:{64,1,1} sharedMem:+0 stream:0.1 @12376144121259.[0m

This patch automatically consolidates contiguous memcpy operations, reducing the number of calls that need to be made. I have workloads where it doubles the training speed.

ROCm-Apps-Test commented:
Can one of the admins verify this patch?

ROCm-Apps-Test commented:
Can one of the admins verify this patch?

ROCm-Apps-Test commented:
Can one of the admins verify this patch?

ROCm-Apps-Test commented:
Can one of the admins verify this patch?

sunway513 commented:
jenkins: ok to test

jerryyin commented on tensorflow/stream_executor/rocm/rocm_blas.cc:
```diff
@@ -1859,18 +1859,48 @@ port::Status ROCMBlas::AllocateStridedBuffer(
 
+  char* src_ptr=0;
+  char* dst_ptr=0;
+  uint64_t cur_size=0;
```
Could you rename this to `cur_stride_size`?

jerryyin commented on tensorflow/stream_executor/rocm/rocm_blas.cc:
```diff
@@ -1859,18 +1859,48 @@ port::Status ROCMBlas::AllocateStridedBuffer(
+      cur_size = matrix_byte_size;
+    }
+  }
+  if(cur_size!=0)
```
Could you add a comment explaining that the last block wraps up the last memcpy under the condition that it is constant stride for the last src_ptr?

ekuznetsov139 commented:
I've addressed some of your concerns.

> zen@zen-ubuntu:~/tensorflow/tensorflow$ git clang-format
> git: 'clang-format' is not a git command. See 'git --help'.
> 

I can't see what's wrong with the test failures, because all Jenkins links give me 404 Not Found errors. I can access the web site, but the pages with job status details don't seem to be there (I tried from inside AMD VPN too and that makes no difference.)

I'm not part of the MLSE organization, this is sort of a hobby project for me. 

The particular workload that brought this to my attention still has substantial overhead - the change only takes it from 1024 memcpy calls per matmul to 128 - but it's tolerable. Right now with this workload I'm seeing ~255 ms / step with RX Vega 64, compared with 175 ms / step doing the exact same thing on 1080 Ti. They are usually fairly close to each other. (Before the patch, I was seeing 700 ms / step).  I'd guesstimate that another 50 ms could be saved if memcpys could be completely avoided. This is, of course, highly dependent on problem parameters (matrix dimensions).

jerryyin commented:
@ekuznetsov139 No worries, I can help take look at the test failures as well as formatting issues. I'll ping you when this is done.

I noticed that you are using ROCm on bare metal. A easier way to work and build is to take advantage of our per-commit docker images [here](https://cloud.docker.com/u/rocm/repository/docker/rocm/tensorflow-autobuilds).

@sunway513 FYI, please keep the PR open. I will work on it after I'm back from vacation.

ekuznetsov139 commented:
Right now the only failing CI build is rocm-xla, test //tensorflow/compiler/xla/tests:array_elementwise_ops_test_gpu, and that one has been failing since September 23 (first failing build 704 "Develop upstream sync 190923")

jerryyin commented:
Closing this PR, as all changes pulled into #665
